### PR TITLE
Helpers to simplify implementing HPC thread parallelism

### PR DESCRIPTION
**Pull request by kbowers-jump**
This uses the same HPC, low latency, high scalability, NUMA aware, deterministic, fine-grained load-balanced, player-coach distribution strategy used by `fd_tpool_exec_all_block`.  On a dual socket 40-core Icelake at 2.3 GHz, the overhead to dispatch and synchronize to a single thread was ~4 ns and overhead to dispatch and synchronize a kernel to 64 threads (0-31 on socket 0, 32-63 on socket 1), was ~2.2 us.

But it requires a lot less boilerplate to use.  Example:
```
... define a computational kernel

FD_FOR_ALL_BEGIN( vec_add, 1L ) {

  double const * restrict x = (double const *)_a0;
  double const * restrict y = (double const *)_a1;
  double       * restrict z = (double       *)_a2;

  for( long i=block_i0; i<block_i1; i++ ) z[i] = x[i] + y[i];

} FD_FOR_ALL_END

... have tpool threads [t0,t1) apply this kernel to the range [i0,i1)

FD_FOR_ALL_3( vec_add, tpool, t0,t1, i0,i1, x,y,z );
```

More features and details can be found in the header documentation.

In short, it's a lot like writing CUDA kernels but targeted at MIMD CPUs and then with a lot less overhead to fire up a kernel (and there no need to implement all the data choreography between the CPU and GPU or eat the overhead to shuffle data back and forth at run time).

### CODE REVIEW

File: src/util/tpool/fd_tpool.h [hunk 1]
```diff
@@ -923,6 +940,277 @@ fd_tpool_exec_all_taskq( fd_tpool_t *    tpool,
...
+
+#define FD_FOR_ALL_BEGIN(op,BLOCK_THRESH)                                                                                         \
+void                                                                                                                              \
+op( void * _tpool,                                                                                                                \
+    ulong   tpool_t0, ulong   tpool_t1,                                                                                           \
+    void * _block_i0, void * _block_i1,                                                                                           \
+    ulong  _a0,       ulong  _a1,                                                                                                 \
+    ulong  _a2,       ulong  _a3,                                                                                                 \
+    ulong  _a4,       ulong  _a5,                                                                                                 \
+    ulong  _a6 ) {                                                                                                                \
```
[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/tpool/fd_tpool.h:1016]: Debated with me quite a bit about this.  The ideal is to write a normal-ish function prototype (e.g. user provided argument types and names with no implicit or explicit casting anywhere) and then do something normal-ish to dispatch and sync it across a target set of threads (e.g. normal call with some annotation of the target set of threads) while keeping performance ultra high, the compile ultra strict and the build ultra fast.

This gets close but it has the argument type and count limitations you noted.

In the current implementation, the function signature needs to be compatible with a fd_tpool_task_t.  The number of arguments is such that the info sent by tpool_exec fits within a cache line pair (e.g. to exploit adjacent cache line prefetch behaviors under the hood) and provides enough flexibility to cover the vast majority of HPC thread dispatch scenarios (e.g. the large dense matrix-matrix multiple example in the HPC threading crash course) while hopefully not creating too many annoying unused arguments in normal circumstances.

This has the side benefit that receiver doesn't need to fetch any additional cache lines after function entry to get the info it needs to start crunching data (e.g. the pthread_create function signature with a single void * argument is terrible in this regard but that's the least concern if trying to do HPC with pthreads).  Additionally, the sent info always arrives in the same place for good cache spatial and temporal locality.  Further, the fixed types allow the packing / unpacking process to be heavily optimized at compile time.

With this in mind, one option is to keep the communication process the same but make argument packing / unpacking more flexible.  A wrinkle here is that compilers are getting much more picky in strict compiles about function pointer casts (even when explicit and perfectly safe according to target ABI).  A second wrinkle is handling arguments that don't fit (should it crash at compile time, should it crash at run time, should it fallback to a degraded performance, is the failure mode plausibly diagnosable ...)  Another wrinkle is that it isn't clear how good a job the compiler does at the more flexible packing, especially if implemented conventionally (e.g. via stdarg).  Etc.

A second option would be to use a more flexible communication process.  But retaining good performance, conformance and portability looked tricky when sketching out designs for this (performance and interoperability between stdarg and thread isn't exactly a top concern by compilers / languages / etc ... e.g. suspect that trying something like passing a va_list between threads would open Pandora's box).

A third option is to be less ambitious and use variadic macros to simplify the call site (no icky _N suffixes).  This feels plausible but I've historically had such bad experiences with cross platform variadic macro support that I didn't want to try it.  (Actually, I was surprised I was able to get this far without needing to write a separate CUDA-style code preprocessing step in the build.)

TL;DR Mixed views but open to improvements.

[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/tpool/fd_tpool.h]: This question (and the cold weather) inspired me to do battle with variadic macros again.  In the just amended commit, the icky _N suffixes on FD_FOR_ALL and FD_MAP_REDUCE are gone (and, if too many or tow few arguments are used to these, it will fail at compile time with an understandable message).

To support this `FD_VA_ARGS_SELECT` was added to util_base.h compiler tricks.  (And, if there was a way to convert a VA_ARGS list of the form "type_t arg, another_type_t another_arg, ..." into "(ulong)(arg), (ulong)another_arg, ..." at compile time, we could get rid argument name and type restrictions too.  But suspect that is beyond the limits of what is possible with standard preprocessor tricks.)

In doing this, realized the thread dispatching/sync logic could be further tweaked to use only one scratch frame and one stack frame, which seems useful.  So also updated that.

In any case, this amend undid your earlier review.  So please rereview at your convenience.

**Review by ripatel-fd — APPROVED** [review]
